src/main/java/com/boot/cart/service/CartService.java

```diff
@@ -1,255 +1,251 @@
+            CartEntry cartEntry = cartEntries.stream().filter(entry -> productName.equals(entry.getProductName())).findFirst().orElse(null);
+
+            if (cartEntry != null) {
+                cartEntry.setQuantity(cartEntry.getQuantity() + quantity);
+                cartEntry.setCart(cart);
+            }else {
+                CartEntry newCartEntry = new CartEntry();
+                newCartEntry.setProductName(productName);
+                newCartEntry.setQuantity(quantity);
+                newCartEntry.setCart(cart);
+                cartEntryRepository.save(newCartEntry);
+            }
         }
 
-        cart.setUser(user);
-
         for (int i = 0; i < quantity; i++) {
-            productList.add(product);
-
-            productTotal += product.getProductPrice();
+            productTotal += productDTO.getPrice();
         }
 
-        cart.setProductList(productList);
+        cart.setEntries(cartEntries);
         cart.setTotal(cart.getTotal() + productTotal);
+        cart.setUserId(user.getId());
 
         cartRepository.save(cart);
 
-        return CartMapper.cartEntityToDto(cart);
-
+        return cartEntityToDto(cartRepository.findByUserId(user.getId()));
     }
 
     public CartDTO updateProductFromCart(String email, String productName, int quantity)
             throws InvalidInputDataException, EntityNotFoundException {
+        log.info("removeProductFromCart - process started");
 
-        Product product;
+        ProductDTO productDTO;
         try {
-            product = ProductMapper.DtoToProductEntity(productServiceClient.callGetProductByProductName(productName));
+            productDTO = productServiceClient.callGetProductByProductName(productName);
         } catch (HttpClientErrorException.NotFound e) {
             throw new EntityNotFoundException("Product: " + productName + " not found in the Database!");
         }
 
-        User user;
+        if (productDTO.getStock() == 0) {
+            throw new InvalidInputDataException("We are sorry, but currently: " + productName + " is out of order!");
+        }
+
+        if (productDTO.getStock() < quantity) {
+            throw new InvalidInputDataException("You can not remove more than: " + productDTO.getStock() + " "
+                    + productName + " Products from your shopping cart!");
+        }
+        UserDTO user;
         try {
-            user = UserMapper.DtoToUserEntity(userServiceClient.callGetUserByEmail(email));
+            user = userServiceClient.callGetUserByEmail(email);
         } catch (HttpClientErrorException.NotFound e) {
-            throw new EntityNotFoundException("UserName: " + email + " not found in the Database!");
+            throw new EntityNotFoundException("Email: " + email + " not found in the Database!");
         }
-        Cart cart = cartRepository.findByUser(user);
-        List<Product> productList = cart.getProductList();
+        Cart cart = cartRepository.findByUserId(user.getId());
```

Comment:
I think you should validate that the cart exists in order to prevent NPE a few lines later

Reply:
done!